primitives/account/src/account/staking_contract/staker.rs

```diff
@@ -500,96 +551,175 @@ impl StakingContract {
-        // We only need to wait for the release time if stake is delegated.
-        if let Some(validator_address) = &staker.delegation {
-            // Fail if the inactive release block height has not passed yet.
-            if let Some(inactive_from) = staker.inactive_from {
-                if block_number < Policy::block_after_reporting_window(inactive_from) {
-                    debug!(
-                        ?staker.address,
-                        "Tried to remove stake while the inactive balance has not been released yet"
-                    );
-                    return Err(AccountError::InvalidForSender);
-                }
-            }
+        tx_logger: &mut TransactionLog,
+    ) -> Result<RetireStakeReceipt, AccountError> {
+        // Get the staker.
+        let mut staker = store.expect_staker(staker_address)?;
 
-            // Fail if validator is currently jailed.
-            if let Some(validator) = store.get_validator(validator_address) {
-                if validator.is_jailed(block_number) {
-                    debug!(
-                        ?staker.address,
-                        "Tried to remove stake that is currently delegated to a jailed validator"
-                    );
-                    return Err(AccountError::InvalidForSender);
-                }
-            }
+        // Fail if staker does not have sufficient funds to retire.
+        if staker.inactive_balance < retire_stake {
+            return Err(AccountError::InsufficientFunds {
+                needed: retire_stake,
+                balance: staker.inactive_balance,
+            });
         }
+
+        // Fail if the minimum stake would be violated for the non retired funds (invariant 1).
+        let new_inactive_balance = staker.inactive_balance - retire_stake;
+        if !Self::is_min_stake_enforced(
+            staker.balance,
+            new_inactive_balance,
+            staker.retired_balance + retire_stake,
+        ) {
+            debug!(
+                active_balance=?staker.balance,
+                inactive_balance=?staker.inactive_balance,
+                retired_balance=?staker.retired_balance,
+                ?retire_stake,
+                "Tried to retire stake that would violate the minimum stake"
+            );
+            return Err(AccountError::InvalidCoinValue);
+        }
+
+        // Fail if the the funds are in cooldown or jailed.
+        if !self.is_inactive_stake_released(store, &staker, block_number) {
+            debug!(
+                ?staker_address,
+                "Tried to retire stake with inactive balance locked (cooldown or jail)"
+            );
+            return Err(AccountError::InvalidForRecipient);
+        }
+
+        // All checks passed, not allowed to fail from here on!
+
+        // Store old values for receipt.
+        let old_inactive_from = staker.inactive_from;
+
+        // Update the staker's balances.
+        staker.inactive_balance -= retire_stake;
```

Comment:
You created this variable above, might as well use it.
```suggestion
        staker.inactive_balance = new_inactive_balance;
```